cpp/src/parquet/encoding.cc

```diff
@@ -2355,6 +2355,80 @@ class DeltaLengthByteArrayDecoder : public DecoderImpl,
   std::shared_ptr<ResizableBuffer> buffered_data_;
 };
 
+// ----------------------------------------------------------------------
+// RLE_BOOLEAN_DECODER
+
+class RleBooleanDecoder : public DecoderImpl, virtual public BooleanDecoder {
+ public:
+  explicit RleBooleanDecoder(const ColumnDescriptor* descr)
+      : DecoderImpl(descr, Encoding::RLE) {}
+
+  void SetData(int num_values, const uint8_t* data, int len) override {
+    num_values_ = num_values;
+    uint32_t num_bytes = 0;
+
+    if (len < 4) {
+      throw ParquetException("Received invalid length : " + std::to_string(len) +
+                             " (corrupt data page?)");
+    }
+    // Load the first 4 bytes in little-endian, which indicates the length
+    num_bytes =
+        ::arrow::bit_util::ToLittleEndian(::arrow::util::SafeLoadAs<uint32_t>(data));
+    if (num_bytes < 0 || num_bytes > (uint32_t)(len - 4)) {
+      throw ParquetException("Received invalid number of bytes : " +
+                             std::to_string(num_bytes) + " (corrupt data page?)");
+    }
+
+    const uint8_t* decoder_data = data + 4;
+    decoder_ = std::make_shared<::arrow::util::RleDecoder>(decoder_data, num_bytes,
+                                                           /*bit_width=*/1);
+  }
+
+  int Decode(bool* buffer, int max_values) override {
+    max_values = std::min(max_values, num_values_);
+    int val = 0;
```

Comment:
`bool`?

Or can we use `decoder_->GetBatch()` in this method too?

Reply:
Changing to `bool` or `decoder_->GetBatch()` fails in compilation stage. 

This barfs at compile time check with 
```
bit_stream_utils.h:415:41: error: no matching function for call to ‘FromLittleEndian(bool&)’
  415 |   *v = arrow::bit_util::FromLittleEndian(*v);

arrow/util/type_traits.h: In substitution of ‘template<class T, class ... Args> using EnableIfIsOneOf = typename std::enable_if<arrow::internal::IsOneOf<T, Args ...>::value, T>::type [with T = bool; Args = {long int, long unsigned int, int, unsigned int, short int, short unsigned int, unsigned char, signed char, float, double}]’:
```


Code path 
**RleDecoder::GetBatch(T*, int) -> RleDecoder::NextCounts() -> BitReader::GetAligned(int, T*)**

In BitReader::GetAligned() :::: Line 415 , there is a call to bit_util::FromLittleEndian. 
However FromLittleEndian can be called on only one of the following types int8...int64, float and double, whereas in this case it is bool All the accepted values are at the byte boundary / 8 bits --- 8,16,32. I am not exactly sure how boolean can fit into this structure. Little Endian or Big Endian wouldn't matter for boolean, but I worry this might change the computation because of the alignment of byte boundary.

Reply:
Hmm. Do you know what value is used for boolean in Parquet? (`0` and `1`?) And do you also know the size of a boolean value in Parquet? (1bit? 8bits?)
I can't find them in https://parquet.apache.org/docs/file-format/data-pages/encodings/ .

I think that we need to use the same type for this. If Parquet uses 8bits, we can use `int8_t` or `uint8_t` here.

Reply:
Parquet uses `0` and `1` for boolean in Parquet (Apart from nulls). 
Size of boolean value in Parquet is 1 bit (Logically).

Boolean is only supported in 2 encoding - [Plain](https://parquet.apache.org/docs/file-format/data-pages/encodings/#a-nameplainaplain-plain--0) and [RLE](https://parquet.apache.org/docs/file-format/data-pages/encodings/#a-namerlearun-length-encoding--bit-packing-hybrid-rle--3). Both of them use some part of bit-packing. Although in RLE cases, we might have lesser size because of continuous number, but physically never greater than 1 bit.

Reply:
Thanks.

I re-read the implementation and understand that `RleDecoder` reads 1 bit in this case and write the read value to `buffer` with cast. (1 bit boolean value -> `static_cast<int>(1)` for `true`/`static_cast<int>(0)` for `false` in this case.) So we can use any type for `val`. But I think that `bool` is better than `int` because we want to return `bool` values.

How about this?

```diff
diff --git a/cpp/src/arrow/util/endian.h b/cpp/src/arrow/util/endian.h
index f77077f809..d428287882 100644
--- a/cpp/src/arrow/util/endian.h
+++ b/cpp/src/arrow/util/endian.h
@@ -122,28 +122,28 @@ static inline void ByteSwap(void* dst, const void* src, int len) {
 #if ARROW_LITTLE_ENDIAN
 template <typename T, typename = internal::EnableIfIsOneOf<
                           T, int64_t, uint64_t, int32_t, uint32_t, int16_t, uint16_t,
-                          uint8_t, int8_t, float, double>>
+                          uint8_t, int8_t, float, double, bool>>
 static inline T ToBigEndian(T value) {
   return ByteSwap(value);
 }
 
 template <typename T, typename = internal::EnableIfIsOneOf<
                           T, int64_t, uint64_t, int32_t, uint32_t, int16_t, uint16_t,
-                          uint8_t, int8_t, float, double>>
+                          uint8_t, int8_t, float, double, bool>>
 static inline T ToLittleEndian(T value) {
   return value;
 }
 #else
 template <typename T, typename = internal::EnableIfIsOneOf<
                           T, int64_t, uint64_t, int32_t, uint32_t, int16_t, uint16_t,
-                          uint8_t, int8_t, float, double>>
+                          uint8_t, int8_t, float, double, bool>>
 static inline T ToBigEndian(T value) {
   return value;
 }
 
 template <typename T, typename = internal::EnableIfIsOneOf<
                           T, int64_t, uint64_t, int32_t, uint32_t, int16_t, uint16_t,
-                          uint8_t, int8_t, float, double>>
+                          uint8_t, int8_t, float, double, bool>>
 static inline T ToLittleEndian(T value) {
   return ByteSwap(value);
 }
@@ -153,28 +153,28 @@ static inline T ToLittleEndian(T value) {
 #if ARROW_LITTLE_ENDIAN
 template <typename T, typename = internal::EnableIfIsOneOf<
                           T, int64_t, uint64_t, int32_t, uint32_t, int16_t, uint16_t,
-                          uint8_t, int8_t, float, double>>
+                          uint8_t, int8_t, float, double, bool>>
 static inline T FromBigEndian(T value) {
   return ByteSwap(value);
 }
 
 template <typename T, typename = internal::EnableIfIsOneOf<
                           T, int64_t, uint64_t, int32_t, uint32_t, int16_t, uint16_t,
-                          uint8_t, int8_t, float, double>>
+                          uint8_t, int8_t, float, double, bool>>
 static inline T FromLittleEndian(T value) {
   return value;
 }
 #else
 template <typename T, typename = internal::EnableIfIsOneOf<
                           T, int64_t, uint64_t, int32_t, uint32_t, int16_t, uint16_t,
-                          uint8_t, int8_t, float, double>>
+                          uint8_t, int8_t, float, double, bool>>
 static inline T FromBigEndian(T value) {
   return value;
 }
 
 template <typename T, typename = internal::EnableIfIsOneOf<
                           T, int64_t, uint64_t, int32_t, uint32_t, int16_t, uint16_t,
-                          uint8_t, int8_t, float, double>>
+                          uint8_t, int8_t, float, double, bool>>
 static inline T FromLittleEndian(T value) {
   return ByteSwap(value);
 }
diff --git a/cpp/src/parquet/encoding.cc b/cpp/src/parquet/encoding.cc
index 98626c9791..b9257aab22 100644
--- a/cpp/src/parquet/encoding.cc
+++ b/cpp/src/parquet/encoding.cc
@@ -2386,18 +2386,8 @@ class RleBooleanDecoder : public DecoderImpl, virtual public BooleanDecoder {
 
   int Decode(bool* buffer, int max_values) override {
     max_values = std::min(max_values, num_values_);
-    int val = 0;
-
-    for (int i = 0; i < max_values; ++i) {
-      if (!decoder_->Get(&val)) {
-        throw ParquetException("Unable to parse bits for position (0 based) : " +
-                               std::to_string(i) + " (corrupt data page?)");
-      }
-      if (val) {
-        buffer[i] = true;
-      } else {
-        buffer[i] = false;
-      }
+    if (decoder_->GetBatch(buffer, max_values) != max_values) {
+      ParquetException::EofException();
     }
     num_values_ -= max_values;
     return max_values;
```

Reply:
Agreed. This is apt and concise. I have changed my implementation as recommended.